### ISSUE TITLE
using bedtools coverage with karyoploteR

### ISSUE THREAD

**Issue by frankbioinfo** [bernatgel/karyoploteR#16]
Hello Team, 

is there a way we can use output of bedtools coverage with karyoploterR 

Specifically this function kpPlotCoverage(kp, data=regions)

many thanks

**Comment by bernatgel** [comment]
Hi @frankbioinfo 

There's no ready-made function in karyoploteR to use bedtools output to plot coverage. 

bedtools has many option and bedtools coverage can create quite different output depending on a couple of parameters. If you want to use the output from bedtools you'll have to load the results into R and use them to create the plot, which may take quite a lot of memory if you are  working on the whole genome. 

If you want to compute the coverage out of R and plot it with karyoploteR, you could use the new kpPlotBigWig function, but it's only available in the devel version of karyoploteR.

What's you main reason to want to plot bedtools output? Do you have a speed or memory problem when using kpPlotCoverage?

**Comment by frankbioinfo** [comment]
Hello @bernatgel 
Thanks for response, I have nimbelgen capture data want to see over all how much coverage it got in each chromosome as compared to regular WGS data  against probes we designed for  I was planning to use  plotKaryotype(plot.type=2) to plot both the data sets for this purpose, may be my understanding in how to use karyoploteR for this purpose is not good any tips.

thanks again

**Comment by bernatgel** [comment]
Hi @frankbioinfo 

If you have BAM files and want to get an overall impression of the coverage, you can use kpPlotBAMDensity. It breaks the whole genome in windows of a given size and counts the number of reads overlapping each window. Windows can be as small as a single base or as large as whole chromosomes. By default, with `normalize=TRUE` it plots the number of reads overlapping each part of the genome normalized by the total of reads in the BAM file, but if you set `normalize=FALSE` it will plot the exact number of reads overlapping the window. 

There's currently large room for improvement in efficiency in this function, specially for small window sizes. If you end up using it and need it to be faster, tell me and I can work on that.

Bernat

**Comment by frankbioinfo** [comment]
Let me rephrase my question I created granges object which looks like this 
```
gr_obj
GRanges object with 2581 ranges and 1 metadata column:
         seqnames              ranges strand |        name
            <Rle>           <IRanges>  <Rle> | <character>
     [1]     chr1       758597-758696      * |      373.07
     [2]     chr1     1017444-1017730      * | 509.0139373
     [3]     chr1     1047659-1047961      * | 1083.858086
     [4]     chr1     1149064-1149309      * | 770.0650407
     [5]     chr1     1174579-1174832      * |  427.980315
     ...      ...                 ...    ... .         ...
  [2577]     chr1 248607558-248607828      * | 1506.077491
  [2578]     chr1 248607833-248607968      * | 2321.514706
  [2579]     chr1 248707330-248707531      * | 141.8910891
  [2580]     chr1 248810573-248810810      * | 928.5672269
  [2581]     chr1 248810833-248811068      * |  1362.84322
-------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
where name is basically mean coverage for that position for trial purposes i am just plotting chromosome 1  when i plot this with 
```
kp <- plotKaryotype(plot.type=2)
kpPlotCoverage(kp, data=gr_obj, data.panel = 1, col="blue")
```

what i see is this coverage across the whole chromosome, which should not be the case. 

![screen shot 2018-08-07 at 11 53 30 am](https://user-images.githubusercontent.com/27443838/43790589-a12b4248-9a38-11e8-9f2a-69d686334537.png)

Any thoughts,

Thanks

**Comment by bernatgel** [comment]
Ok, it makes total sense. [kpPlotCoverage](https://bernatgel.github.io/karyoploter_tutorial//Tutorial/PlotCoverage/PlotCoverage.html) does plot the number of elements in the GRanges object on every base of the genome. In this case, it's either 0 or 1, so you see this in the plot. This function cannot be used to plot the coverage of a BAM file if it has already been computed somewhere else. To do that, you can use any of the "basic" plotting functions of the package. For your example, where you assign a coverage value to regions more than 1 base wide, I think [kpBars](https://bernatgel.github.io/karyoploter_tutorial//Tutorial/Bars/Bars.html) would be the best option.

For example, creating an example similar to yours:

```R
  regs
  GRanges object with 339 ranges and 1 metadata column:
          seqnames            ranges strand |             name
             <Rle>         <IRanges>  <Rle> |        <numeric>
      [1]     chr1 14961328-15961327      * | 900.922351633199
      [2]     chr1 19271785-20271784      * | 611.219123494811
      [3]     chr1 23915767-24915766      * | 354.461921099573
      [4]     chr1 25340724-26340723      * | 880.639736633748
      [5]     chr1 34530758-36228692      * | 861.533168866299
      ...      ...               ...    ... .              ...
    [335]     chrY 19375883-20375882      * | 190.574221429415
    [336]     chrY 27056686-28056685      * | 793.772204429843
    [337]     chrY 29924018-31392781      * | 756.549454247579
    [338]     chrY 45554053-46554052      * | 747.502454393543
    [339]     chrY 51231336-52231335      * |  801.78132141009
    -------
    seqinfo: 24 sequences from an unspecified genome; no seqlengths
```
To plot kpBars, we need a column called `y1` or we have to specify the height of the bar in the `y1`parameter. In addition, the default maximum value in a plot is 1, so since we have coverage values greater than 1, we'll need to set `ymax` to a suitable value. In this case, I set it to the maximum of the coverage values. 

```R
kp <- plotKaryotype()
kpBars(kp, regs,  y1=regs$name, ymax=max(regs$name), col="blue", border=NA)
```

and this will create a plot like this, which I think it's what you were expecting

![example_with_kpBars](https://user-images.githubusercontent.com/919489/43820331-3f110c6a-9ae6-11e8-9ebc-a03fea3c36c5.png)

Does it help?

**Comment by frankbioinfo** [comment]
@bernatgel thanks Yes this helps, would you know how to increase height of those peaks in my case i do not see such a clear , also i do not see the colors 

```
kp <- plotKaryotype(plot.type = 2)
kpBars(kp, gr_obj,data.panel = 1,y1=gr_obj$name, ymax = max(gr_obj$name), col="blue")
kpBars(kp, data=gr_obj1, data.panel = 2, y1=gr_obj1$name, ymax=max(gr_obj1$name), col="red")
```

<img width="933" alt="screen shot 2018-08-10 at 9 38 10 am" src="https://user-images.githubusercontent.com/27443838/43963903-450d7e90-9c81-11e8-8ab2-348ed315ea63.png">

**Comment by bernatgel** [comment]
To increase the aparent size of the bars all you need to do is lower ymax, if you set ymax to "max(gr_obj1$name)/2", your bars will be twice as hight.

Since you are plotting tiny bars the black border is the only visible part. Set "border=NA" in the kpBars call to disable the border. 

Oh, and I would recommend creating plot with "zoom" to show smaller parts of the genome, one at a time, where you can actually make sense of the data!
